### PR TITLE
Add shims for EcKey, EvpPkey, and EcDsa files

### PR DESCRIPTION
**Pull request by eerhardt**
Add shims for all the remaining libcrypto calls, except Interop.Rsa.cs.

@bartonjs, @nguerrera, @sokket

### CODE REVIEW

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs [hunk 1]
```diff
@@ -245,10 +245,7 @@ public RSA GetRSAPrivateKey()
                 return null;
             }
 
-            using (SafeRsaHandle rsaHandle = Interop.libcrypto.EVP_PKEY_get1_RSA(_privateKey))
```
Comment (eerhardt) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs]: @bartonjs - are the two changes to this file OK? Doing it this way allowed me to not have to include a bunch of interop files in the X509Certificates.csproj. My naïve thinking is that this approach is a little different than existing code, but should be fine.  The current code calls EVP_PKEY_get1_* here and then passes in the IntPtr into the constructor, which then just duplicates the handle and then this code disposes of the first handle.  The new way is to pass in the privateKey into the constructor, which will call EVP_PKEY_get1_* and just hold on to that handle.  So this way seems simpler/more performant to me.

Thoughts?

**Comment by stephentoub** [comment]
LGTM

**Comment by sokket** [comment]
LGTM 2

**Comment by nguerrera** [comment]
LGTM

File: src/Native/System.Security.Cryptography.Native/pal_evp_pkey.h [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+The given EVP_PKEY pointer is invalid after this call.
+Always succeeds.
+*/
+extern "C" void EvpPkeyDestroy(EVP_PKEY* pkey);
+
+/*
+Function:
+UpRefEvpPkey
+
+Used by System.Security.Cryptography.X509Certificates' OpenSslX509CertificateReader when
```
Comment (bartonjs) [inline src/Native/System.Security.Cryptography.Native/pal_evp_pkey.h]: I'd alter these to match the style of the rest of the shims; and get rid of the CoreClrPal notion of who cares about it.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs (hunk 1 above)
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509CertificateReader.cs]: Seems reasonable.  When this code was written there wasn't an SafeEvpPKeyHandle constructor :)

**Comment by bartonjs** [comment]
One comment about a comment, otherwise LGTM.

File: src/Native/System.Security.Cryptography.Native/pal_ecdsa.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+{
+    if (!siglen)
+    {
+        return 0;
+    }
+
+    unsigned int unsignedSigLength = UnsignedCast(*siglen);
+    int ret = ECDSA_sign(0, dgst, dgstlen, sig, &unsignedSigLength, key);
+    assert(unsignedSigLength <= INT32_MAX);
+    *siglen = static_cast<int32_t>(unsignedSigLength);
```
Comment (nguerrera) [inline src/Native/System.Security.Cryptography.Native/pal_ecdsa.cpp]: We should add SignedCast for this.

File: src/Native/System.Security.Cryptography.Native/pal_evp_pkey.h (hunk 2 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_evp_pkey.h]: Cleaned up.

File: src/Native/System.Security.Cryptography.Native/pal_ecdsa.cpp (hunk 3 above)
Comment (eerhardt) [inline src/Native/System.Security.Cryptography.Native/pal_ecdsa.cpp]: Done.

**Pull request merged**